daemon/image.go

```diff
@@ -113,12 +121,233 @@ func (daemon *Daemon) GetCachedImage(imgID image.ID, config *containertypes.Conf
+		} else {
+			historyWithSource[i].sourceLayerID = image.RootFS.DiffIDs[layerIndex]
+		}
+
+		// now increment, if needed, and look at the result layer ID
+		if !h.EmptyLayer {
+			layerIndex = layerIndex + 1
+		}
+		if layerIndex == -1 {
+			historyWithSource[i].resultingLayerID = digest.DigestSha256EmptyTar
+		} else {
+			historyWithSource[i].resultingLayerID = image.RootFS.DiffIDs[layerIndex]
+		}
+
+		// Copy the other history entries over I'm interested in
+		historyWithSource[i].cmd = h.CreatedBy
+		historyWithSource[i].createdAt = h.Created
+		historyWithSource[i].history = h
+	}
+
+	return historyWithSource
+}
+
 // GetCachedImageOnBuild returns a reference to a cached image whose parent equals `parent`
 // and runconfig equals `cfg`. A cache miss is expected to return an empty ID and a nil error.
-func (daemon *Daemon) GetCachedImageOnBuild(imgID string, cfg *containertypes.Config) (string, error) {
-	cache, err := daemon.GetCachedImage(image.ID(imgID), cfg)
-	if cache == nil || err != nil {
+func (cache *daemonImageCacheForBuild) GetCachedImageOnBuild(imgID string, cfg *containertypes.Config) (string, error) {
+	cachedImage, err := cache.daemon.GetCachedImage(image.ID(imgID), cfg)
+	if err != nil {
 		return "", err
 	}
-	return cache.ID().String(), nil
+	if cachedImage != nil {
+		// We found a cache hit using the old parent image method
+		return cachedImage.ID().String(), nil
+	}
+
+	// We didn't find a cache hit using that method. Explore cacheFrom images for matching history
+	var parentImage *image.Image
+	var parentImageHistory []historyWithSourceT
+	// i.e. not FROM SCRATCH
+	if imgID != "" {
+		parentImage, err = cache.daemon.imageStore.Get(image.ID(imgID))
+		if err != nil {
+			return "", err
+		}
+		parentImageHistory = makeHistoryWithSource(parentImage)
+	}
+
+	// For each thing we are caching from, see if it matches parentImageHistory
+	type matchStruct struct {
+		cacheFrom   string
+		nextLayerID layer.DiffID
+		createdAt   time.Time
+		history     image.History
+	}
+	matches := make([]matchStruct, 0, len(cache.cacheFromImages))
+	for cacheFromName, cacheFromImageHistory := range cache.cacheFromImageHistories {
+
+		if historiesMatch(cacheFromImageHistory, parentImageHistory) &&
+			// Not only do the histories need to match, but cacheFromImageHistory needs to be long enough
+			// that we can peek at the next piece of "history" (which is in the future for parentImage -
+			// that's why we can cache from it!)
+			(len(cacheFromImageHistory) > len(parentImageHistory)) &&
+			// And the command needs to match too.
+			(strings.Join(cfg.Cmd, " ") == cacheFromImageHistory[len(parentImageHistory)].cmd) {
+
+			match := matchStruct{
+				cacheFrom:   cacheFromName,
+				nextLayerID: cacheFromImageHistory[len(parentImageHistory)].resultingLayerID,
+				createdAt:   cacheFromImageHistory[len(parentImageHistory)].createdAt,
+				history:     cacheFromImageHistory[len(parentImageHistory)].history,
```

Comment:
`cacheFromImageHistory[len(parentImageHistory)]` appears several times since line 251. It would clean things up a little bit to use an intermediate variable.